qadence/ml_tools/saveload.py

```diff
@@ -41,19 +46,24 @@ def load_checkpoint(
     optimizer: Optimizer | NGOptimizer,
     model_ckpt_name: str | Path = "",
     opt_ckpt_name: str | Path = "",
+    device: str | torch.device = "cpu",
 ) -> tuple[Module, Optimizer | NGOptimizer, int]:
     if isinstance(folder, str):
         folder = Path(folder)
     if not folder.exists():
         folder.mkdir(parents=True)
         return model, optimizer, 0
-    model, iter = load_model(folder, model, model_ckpt_name)
-    optimizer = load_optimizer(folder, optimizer, opt_ckpt_name)
+    model, iter = load_model(folder, model, model_ckpt_name, device)
+    optimizer = load_optimizer(folder, optimizer, opt_ckpt_name, device)
     return model, optimizer, iter
 
 
 def write_checkpoint(
-    folder: Path, model: Module, optimizer: Optimizer | NGOptimizer, iteration: int | str
+    folder: Path,
+    model: Module,
+    optimizer: Optimizer | NGOptimizer,
+    iteration: int | str,
+    device: str | torch.device = "cpu",
```

Comment:
The rest of the updates are great! There’s still an issue in the `write_checkpoint()` function though: it sets `device` to `None` at the start, ignoring both the provided `device` parameter and the default “cpu”.

Due to the try-except structure, failure to retrieve the device from the pyqtorch native circuit goes unnoticed by the user, leading to misleading training logs:

```
INFO  2024-06-27 11:12:44 - qadence.ml_tools.saveload: Writing PINN checkpoint model_PINN_ckpt_199_device_None.pt
INFO  2024-06-27 11:12:44 - qadence.ml_tools.saveload: Writing SGD to checkpoint opt_SGD_ckpt_199_device_None.pt
```

To address this, I would consider:

1. Omitting `device=None`, allowing the function to use the `device` specified in the input if the pyqtorch native circuit extraction does not succeed. Note: this change would necessitate providing a `device` to the  `write_checkpoint` calls in `train()` to prevent defaulting to “cpu” when a GPU is in use.
2. Keeping the current setup but adding an error notification to alert the user when `device` is mistakenly set to `None`